### PR TITLE
 Desktop: Create a screen that list all the resouces (#592)

### PR DESCRIPTION
**Pull request by talkdirty**
This addresses #592. Adds a new screen under Options, titled "Resources":

DeepinScreenshot_select-area_20191215020222

The resources screen has the following capabilities

* Listing all Resources, including title, size, id
* Display Orphan status
* Ability to delete a Resource

DeepinScreenshot_select-area_20191215020153

This is my first attempt to contribute to Joplin, so feel free to criticize! A few remarks:

* I'm not sure I've fully grasped how to detect the Orphan status yet. Please, someone who knows their way around check the `isOrphan` function before merging.
* Is it okay to do queries from the React component like this or should this be extracted into service / lib layer?
* I saw somewhere new code should be added as Typescript, so I did that. What's the overall plan on typescript migration? For example, if I touch something like `lib/model/Resources`, should I add typescript definitions for the model? Also, do we check in the compiled JS (I saw the compiled JS was checked in before, so I did that also)

Thanks!

### CODE REVIEW

**Comment by laurent22** [comment]
Thanks, that looks quite good already. I need to review the code but just a few comments for now:

- There's no reliable way to know if a resource is an orphan or not so perhaps we shouldn't display this info as that might make people delete resources that they shouldn't?

- Business logic should indeed go in model classes (for db stuff) or services (for me general logic that uses multiple models, does data processing, etc), and be unit tested.

- how well does the table scale? Some users have like millions of resources (often small images or icons from web clipped pages). Would the page still when in this case?

- TypeScript is new here so there's no precise guideline at the moment. I don't think we'll have definition files though. If you want to use an existing js file from a TypeScript file that's fine, you just make it work using untyped code I guess and using `any`.

**Comment by tessus** [comment]
@laurent22 
> There's no reliable way to know if a resource is an orphan or not so perhaps we shouldn't display this info as that might make people delete resources that they shouldn't?

That's actually a good point. However, I think the statement I use in my script is able to ascertain, if a resource is an orphan. (Unless I really screwed it up.) It takes the note history into account as well.

```
select resource_id from note_resources where is_associated = 0 and resource_id not in (select resource_id from note_resources where is_associated = 1) group by resource_id having max(last_seen_time) < strftime('%s','now','-${KEEP} days')*1000

KEEP is the `Keep note history for (days)` value
```

**Comment by laurent22** [comment]
> That's actually a good point. However, I think the statement I use in my script is able to ascertain, if a resource is an orphan. (Unless I really screwed it up.) It takes the note history into account as well.

I think your script gives a good indication of what might be an orphan resource, but there's still the edge case of a resource being created, then deleted right away, without having been indexed. In that case, I think there's no way to know if it's an orphan resources or one that's been synced, but the associated note hasn't been downloaded yet.

**Comment by tessus** [comment]
> but there's still the edge case of a resource being created, then deleted right away, without having been indexed. In that case, I think there's no way to know if it's an orphan resources or one that's been synced, but the associated note hasn't been downloaded yet.

Yes, you are right. But we can minimize the risk by only showing an "orphaned" status after a sync is complete. Won't that work? (e.g. if one opens that window during a sync - don't disply the orphaned status.)
On one side, I don't think we really need an orphaned status in the list, but if there's a way, why not? If it's too complicated, let's drop it.

**Comment by laurent22** [comment]
> But we can minimize the risk by only showing an "orphaned" status after a sync is complete.

Like you say it would minimize the risk, but not remove it. It's still possible that a client has the associated note but hasn't synced it yet.

I think the Orphan status would be good if users understand that it's only a hint, not a certainty. It's a good way to sort the resources and delete those that you know are eno longer needed.

But it's used by some users to batch delete everything marked as Orphan, we'll get lots of support requests in the forum :-)

So I'm not sure how to best handle this. Maybe a big warning message box that explains everything when the user clicks on "delete"?

**Comment by tessus** [comment]
> I think the Orphan status would be good if users understand that it's only a hint, not a certainty. It's a good way to sort the resources and delete those that you know are eno longer needed.

Yep, this would be great.

> But it's used by some users to batch delete everything marked as Orphan, we'll get lots of support requests in the forum :-)

So true.

> So I'm not sure how to best handle this. Maybe a big warning message box that explains everything when the user clicks on "delete"?

Either that or just a warning text above the table, something similar to what you mentioned earlier. "hint, not certainty, be careful, blah - don't blame us"

**Comment by talkdirty** [comment]
Interesting insights. Regarding the Orphans, yes I was noticing already when testing the logic is flaky and indeed might mislead users to trust this (and how blame them given the current state of the changeset). @tessus ' solution looks more robust then the one that is currently in place, which basically will say the resource is an Orphan if, when grouping by the `resource_id` in `note_resources`, all entries' `is_associated = 0`.

However I think I agree that the best thing here is just to drop the Orphan status, maybe keeping it only for sorting, or maybe dropping even that. So the main actual use-case I think there is for this Resource Screen is users wanting to clean up large files that they otherwise can't easily identify inside Joplin. For that use-case it'd be enough to sort by size and be done with it. The other thing that would be really awesome from a user-perspective would be to also show which Notes the resource is referenced in, but from my limited insight into the Joplin codebase it looks like there is no easy way to get that information, apart from using `SearchEngine`, which depends on the search index to be built.

Perhaps more importantly I think we could put in a warning at the top that says something like **Don't worry about Orphaned Resources deletion, Joplin will do this for you automatically**. I think that's an important thing to say because I didn't actually know Joplin did this before looking at the Code - many Users might make the same assumption.

Regarding @laurent22 's valid scalability concerns, I agree we should probably do something to address that. So I'm not actually using Joplin since that long so I only have like 150 Resources, mostly tiny clipper type things. Could someone with lots of resources try what happens from an empirically? That said I think an easy way to address this is just getting a maximum N results, perhaps like 10k, and putting up a warning if the limit is hit. That should be easily manageable by both SQLite and React.

**Comment by tessus** [comment]
Unfortunately I only have 93 resources.

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+const { bridge } = require('electron').remote.require('./bridge');
+const { Header } = require('./Header.min.js');
+const prettyBytes = require('pretty-bytes');
+const Resource = require('lib/models/Resource.js');
+const NoteResource = require('lib/models/NoteResource.js');
+
+interface IProps {
+	style: any;
+	theme: any;
+}
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: Please don't use Hungarian notation. So just `Props`, and same for the other interfaces.

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+		this.state = {
+			resources: undefined,
+			sortingOrders: ['size', 'orphaned'],
+		};
+	}
+
+	async reloadResources() {
+		const noteResources = await NoteResource.modelSelectAll('SELECT * FROM note_resources');
+		const resources = await Resource.modelSelectAll('SELECT * FROM resources');
+		this.setState({ resources: buildResources(resources, noteResources) });
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: If you're going to load all the resources, which maybe is fine, you need at least to select only the columns that you need, such as id, title, etc.

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+	componentDidMount() {
+		this.reloadResources();
+	}
+
+	render() {
+		const style = this.props.style;
+		const theme = themeStyle(this.props.theme);
+		const headerStyle = Object.assign({}, theme.headerStyle, { width: style.width });
+		return <>
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: No empty tag please. Make it a div then.

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+					onResourceDelete={ (resource) => {
+						deleteResource(resource)
+							.catch(error => {
+								bridge().showErrorMessageBox(_('Resource deletion failed: %s', error));
+							})
+							.finally(() => {
+								this.reloadResources();
+							});
+					} }
+				/>}
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: I would prefer if all the event handlers were in separate class methods. That will also mean less unnecessary renderings.

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+const openResource = (resource: IResource) => {
+	const resourcePath = Resource.fullPath(resource.daoResource);
+	const ok = bridge().openExternal(`file://${resourcePath}`);
+	if (!ok) {
+		bridge().showErrorMessageBox(_('This file could not be opened: %s', resourcePath));
+	}
+};
+
+const deleteResource = (resource: IResource): Promise<any> => Resource.batchDelete([resource.id]);
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: Just `Resource.delete()` should work, without an additional wrapper.

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+const buildResources = (resources: any[], noteResources: any[]) => {
+	const groupedNoteResources = groupBy(noteResources, 'resource_id');
+	return resources.map((resource: any) => ({
+		title: resource.title,
+		id: resource.id,
+		isOrphan: isOrphan(groupedNoteResources[resource.id]),
+		daoResource: resource,
+		size: prettyBytes(resource.size),
+	} as IResource));
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: Likewise, if you are going to load all the resources, then perhaps you shouldn't also filter them all like this, as that might be slow. Data formatting ideally should be done when rendering the table but internally we should just keep the regular resource properties.

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 7]
```diff
@@ -0,0 +1,180 @@
+import * as React from 'react';
+import { groupBy, orderBy } from 'lodash';
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: Can we not use lodash? As I'd like to remove this package eventually. If necessary the required functions could be added to ArrayUtils or ObjectUtils.

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+	const sortOrderEngagedMarker = (s: SortingOrder) => (
+		<a href="#" onClick={ () => props.onToggleSortOrder(s) }>{ props.sortingOrders.includes(s) ? '▾' : '▴' }</a>
+	);
+	return <table style={{ width: '90%' }}>
+		<thead>
+			<tr>
+				<th>Title { sortOrderEngagedMarker('name') }</th>
+				<th>Size { sortOrderEngagedMarker('size') }</th>
+				<th>ID</th>
+				<th>Action</th>
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: All this should be localized.

**Comment by laurent22** [comment]
> Could someone with lots of resources try what happens empirically? That said I think an easy way to address this is just getting a maximum N results, perhaps like 10k, and putting up a warning if the limit is hit. That should be easily manageable by both SQLite and React.

Yes just loading n results only would be fine as long as it's possible to sort them and filter them (so that users can get to whatever resource they are looking for). Now you load them all and do the sorting at the front-end level, so I guess you would need to change this so that sorting is done at the SQL level.

**Comment by laurent22** [comment]
By the way you have some eslint issues, so it would be best to install the post commit hook by running `npm i` from the root of the project.

**Comment by tessus** [comment]
@talkdirty what's the status of this?

**Comment by talkdirty** [comment]
@tessus I'll have some time next weekend to work on Laurent's comments and implement the remaining stuff we discussed.

File: ElectronClient/app/gui/ResourceScreen.tsx (hunk 3 above)
Comment (talkdirty) [reply ElectronClient/app/gui/ResourceScreen.tsx]: It's a shorthand for React.Fragment, so it prevents an unnecessary DOM element to be rendered to the real DOM. But sure I'll change to div.

File: ElectronClient/app/gui/ResourceScreen.tsx (hunk 4 above)
Comment (talkdirty) [reply ElectronClient/app/gui/ResourceScreen.tsx]: Sure. I never heard before that putting it as a class method reduced re-renderings, do you happen to have a documentation for that?

File: ElectronClient/app/gui/ResourceScreen.tsx (hunk 7 above)
Comment (talkdirty) [reply ElectronClient/app/gui/ResourceScreen.tsx]: Just for my information, what's the intent behind removing lodash? To me it seems like an excellent collection of frequently used utilities. We can put them into Utils classes, but then we have to maintain them ourselves

Comment (talkdirty) [reply ElectronClient/app/gui/ResourceScreen.tsx]: Sorting after multiple keys (for example size & name), as is possible with the current implementation, is certainly possible with Vanilla JS / some code block copied from somewhere, but I believe letting lodash handle the details is a bit easier to understand. But your call in the end!

Comment (laurent22) [reply ElectronClient/app/gui/ResourceScreen.tsx]: Hmm, I don't know. I remember last time I checked that lodash was quite huge in node_modules, all for relatively small utility functions, but maybe I'm mistaken. Let's leave it as it is for now then, and I'll review later.

**Comment by laurent22** [comment]
> What I could do for now is limiting the results to 10k and keep the frontend sorting. This should be reasonably fast still. If we get significant issues we could still go for creating indices for the things we'd like to sort. However my gut feeling says people with 10k+ resources is an edge case that might not be worth overengineering for (I might be wrong!)

The issue is that if we limit to 10k resources without sorting or filtering at SQL-level, users will always see the same resources without any way to get to the one they need. I know that 10k resources seem huge but it can happen for heavy users of the web clipper, and it's precisely users with a ton of resources that might need this resource table.

If needed we can easily add indexes by adding a migration in `joplin-database.js` (have a look for example at `targetVersion == 6`). To add a migration, you can add a version number to `existingDatabaseVersions`, then add the conditional statement in the loop below. Please let me know what you think.

**Comment by talkdirty** [comment]
Thanks @laurent22 . No I agree, it's a good point we'll always only have the first 10k resources listed, which definitely isn't a scalable solution. I'll look into a database solution instead. Are you worried at all about the migration taking a potentially long time? (e.g. User with 100k Resources)

**Comment by talkdirty** [comment]
Regarding the travis failure - it appears that eslint is checking the transpiled JS that is generated by TS, which is odd. Do I have to check in the JS and run the linter on it? It seems to be done that way in ShareNoteDialog.js.

**Comment by laurent22** [comment]
> Are you worried at all about the migration taking a potentially long time

It's fine as it's a one off. We could benchmark to be sure I guess but SQLite can deal with a lot of data easily. I expect adding an index on 10,000+ row is almost nothing for SQLite.

**Comment by laurent22** [comment]
> Regarding the travis failure - it appears that eslint is checking the transpiled JS that is generated by TS,

I believe it's fixed in latest master?

**Comment by talkdirty** [comment]
OK, I made one more revision which relies on DB sorting (also means we can get rid of lodash in there)
Regarding CI, can you double check? Revision d6fad only has the tsx checked into git but travis fails at `ResourceScreen.js`

**Comment by tessus** [comment]
@marcushill ah, the CI error came from somewhere else. apparently you already added the generated file to `.eslintignore` and `.gitignore`. I think we are good to go.

**Comment by marcushill** [comment]
Me or @talkdirty? I haven't finished the refactor @laurent asked for on my PR.

**Comment by tessus** [comment]
@marcushill sorry, I meant @talkdirty . Not sure how that happened.

**Comment by tessus** [comment]
@talkdirty ah, the CI error came from somewhere else. apparently you already added the generated file to `.eslintignore` and `.gitignore`. I think we are good to go.

**Comment by talkdirty** [comment]
Thanks - let me know when you'd like to merge and I'll resolve the package-lock conflict before.

**Comment by tessus** [comment]
Hmm, somehow 89 files changed doesn't look right. Not sure, if Laurent has already done the final review. But resolving the conflicts can always be done. You don't have to wait for anything.

**Comment by tessus** [comment]
@talkdirty I'm not sure what you have done but this PR is pretty much unusable now. I don't think we can merge this. 203 commits and 388 changed files. ???????

**Comment by laurent22** [comment]
I see 388 changed files even, so it's not possible to review at the moment. I think you need to rebase or something so that unrelated commits don't show up in your branch.

**Comment by talkdirty** [comment]
Yes - sorry the PR branch was rebased on an old version of the master branch. Fixed now.

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+			},
+			isLoading: false,
+		};
+	}
+
+	async reloadResources(sorting: ActiveSorting) {
+		this.setState({ isLoading: true });
+		const resources = await Resource.modelSelectAll(
+			`SELECT title, id, size, file_extension FROM resources ORDER BY ${
+				getSortingOrderColumn(sorting.order)} ${sorting.type} LIMIT ${MAX_RESOURCES}`);
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: Please move the SQL to models/Resource

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+	onResourceDelete(resource: Resource) {
+		Resource.batchDelete([resource.id])
+			.catch((error: Error) => {
+				bridge().showErrorMessageBox(_('Resource deletion failed: %s', error));
+			})
+			.finally(() => {
+				this.reloadResources(this.state.sorting);
+			});
+	}
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: Please use async/await. Also I think `Resource.delete(resource.id)` should work too?

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+	render() {
+		const style = this.props.style;
+		const theme = themeStyle(this.props.theme);
+		const headerStyle = Object.assign({}, theme.headerStyle, { width: style.width });
+		return <div>
+			<Header style={headerStyle} />
+			<div style={{ ...style, margin: '20px', overflow: 'scroll' }}>
+				{this.state.isLoading && <div>{_('Please wait')}...</div>}
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: => `_('Please wait...')` (three dots inside)

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+				{!this.state.isLoading &&<div>
+					{!this.state.resources && <div>
+						{_('No Resources!')}
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: `No resources!` (lowercase)

File: ElectronClient/app/gui/ResourceScreen.tsx (hunk 9 above)
Comment (laurent22) [reply ElectronClient/app/gui/ResourceScreen.tsx]: That should work actually, without modifying models/Resource:

```
const resources = await Resource.all({
	order: [{
		by: getSortingOrderColumn(sorting.order),
		dir: sorting.type,
	}],
	limit: MAX_RESOURCES,
	fields: ['title', 'id', 'size', 'file_extension'],
})
```

Not tested but see BaseModel.all and BaseModel.applySqlOptions or more information.

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	componentDidMount() {
+		this.reloadResources(this.state.sorting);
+	}
+
+	onResourceDelete(resource: Resource) {
+		Resource.batchDelete([resource.id])
+			.catch((error: Error) => {
+				bridge().showErrorMessageBox(_('Resource deletion failed: %s', error));
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: This failure should be very rare, so let's not add a translatable string for it. Just `bridge().showErrorMessageBox(error.message);`

File: ElectronClient/app/gui/ResourceScreen.tsx [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+			.finally(() => {
+				this.reloadResources(this.state.sorting);
+			});
+	}
+
+	openResource(resource: Resource) {
+		const resourcePath = Resource.fullPath(resource);
+		const ok = bridge().openExternal(`file://${resourcePath}`);
+		if (!ok) {
+			bridge().showErrorMessageBox(_('This file could not be opened: %s', resourcePath));
```
Comment (laurent22) [inline ElectronClient/app/gui/ResourceScreen.tsx]: Same here - just `bridge().showErrorMessageBox(error.message);`

**Comment by talkdirty** [comment]
Thanks for bearing with me @laurent22 and @tessus . Addressed the comments in the latest commit.

**Comment by talkdirty** [comment]
Hi @laurent22 @tessus. Anything else that needs doing on this?

**Comment by tessus** [comment]
No, other than the fact that there are conflicting files again, I think we are good.

@laurent22 can we merge this as soon as the conflicts are resolved? I am not sure, since I don't know, if you have reviewed the latest changes.

**Comment by laurent22** [comment]
Ok all good now, thanks for the pull request @talkdirty!

**Pull request merged**